onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlOperatorAttention.cpp

```diff
@@ -0,0 +1,423 @@
+    Cast               Gemm
+     |                / |   \
+     |               /  |    \
+     |              /   |     \
+     |          Slice  Slice  Slice
+  Identity        |     |       |
+     |             -----        |
+     -----------    |           |
+                 \  |           |
+                  Gemm          |
+                    |           |
+                    |           |
+                Softmax         |
+                    |          /
+                    |         /
+                     \       /
+                       \    /
+                        Gemm
+                          |
+                   ActivationLinear
+                          |
+                        Output  // Final output
+
+ This kernel creates a DML_GRAPH, as mentioned above.
+ For refernce refer to this Doc: 
+ https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md#com.microsoft.Attention
+ */
+namespace Dml
+{
+class DmlOperatorAttention : public DmlOperator
+{
+public:
+    DmlOperatorAttention(const MLOperatorKernelCreationContext& kernelCreationContext)
+    :   DmlOperator(kernelCreationContext)
+    {
+        ML_CHECK_VALID_ARGUMENT(kernelCreationContext.GetInputCount() >= 3);
+        DmlOperator::Initialize(kernelCreationContext, std::nullopt, std::nullopt, std::nullopt, std::nullopt, 1);
+
+        std::vector<uint32_t> inputTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(0);
+        std::vector<uint32_t> weightTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(1);
+        std::vector<uint32_t> biasTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(2);
+        std::vector<uint32_t> maskIndexTensorShape = kernelCreationContext.GetTensorShapeDescription().GetInputTensorShape(3);
+        ML_CHECK_VALID_ARGUMENT(inputTensorShape.size() == 3);
+        ML_CHECK_VALID_ARGUMENT(weightTensorShape.size() == 2);
+        ML_CHECK_VALID_ARGUMENT(biasTensorShape.size() == 1);
+        ML_CHECK_VALID_ARGUMENT(weightTensorShape[1] == biasTensorShape[0]);
+        ML_CHECK_VALID_ARGUMENT(biasTensorShape[0] % 3 == 0);
+        ML_CHECK_VALID_ARGUMENT(inputTensorShape[2] == weightTensorShape[0]);
+        // TODO: fix Attention kernel when maskIndexTensorShape is 1
+        // https://microsoft.visualstudio.com/OS/_workitems/edit/41893987
+        ML_CHECK_VALID_ARGUMENT(maskIndexTensorShape.size() > 1 && maskIndexTensorShape.size() <= 4);
+        const uint32_t batchSize = inputTensorShape[0];
+        const uint32_t sequenceLength = inputTensorShape[1];
+        const uint32_t hiddenSize = biasTensorShape[0] / 3;
+        const uint32_t numHeads = gsl::narrow_cast<uint32_t>(kernelCreationContext.GetAttribute<int64_t>(AttrName::NumHeads));
+        ML_CHECK_VALID_ARGUMENT(hiddenSize % numHeads == 0);
+        ML_CHECK_VALID_ARGUMENT(numHeads > 0); // to avoid process crash because of division by zero.
```

Comment:
>ML_CHECK_VALID_ARGUMENT(numHeads > 0); // to avoid process crash because of division by zero. [](http://example.com/codeflow?start=8&length=93)

The division check needs to be before the division, including line 68 because `%` is a division too.

Reply:
(I just made this change and pushed it, since it was faster than force another round of review.)